src/extension/infer.rs

```diff
@@ -663,31 +680,85 @@ impl UnificationContext {
     /// us to come up with a fully concrete solution to pass into validation.
+    ///
+    /// Nodes which loop into themselves must be considered as a "minimum" set
+    /// of requirements. If we have
+    ///   1 = 2 + X, ...
+    ///   2 = 1 + x, ...
+    /// then 1 and 2 both definitely contain X, even if we don't know what else.
+    /// So instead of instantiating to the empty set, we'll instantiate to `{X}`
     pub fn instantiate_variables(&mut self) {
-        for m in self.variables.clone().into_iter() {
+        // A directed graph to keep track of `Plus` constraint relationships
+        let mut relations = GraphContainer::new_directed();
+        let mut solutions: HashMap<Meta, ExtensionSet> = HashMap::new();
+
+        let variable_scope = self.search_variable_deps();
+        for m in variable_scope.into_iter() {
             if !self.solved.contains_key(&m) {
                 // Handle the case where the constraints for `m` contain a self
                 // reference, i.e. "m = Plus(E, m)", in which case the variable
                 // should be instantiated to E rather than the empty set.
-                let solution =
-                    ExtensionSet::from_iter(self.get_constraints(&m).unwrap().iter().filter_map(
-                        |c| match c {
-                            // If `m` has been merged, [`self.variables`] entry
-                            // will have already been updated to the merged
-                            // value by [`self.merge_equal_metas`] so we don't
-                            // need to worry about resolving it.
-                            Constraint::Plus(x, other_m) if m == self.resolve(*other_m) => {
-                                Some(x.clone())
-                            }
-                            _ => None,
-                        },
-                    ));
-                self.add_solution(m, solution);
+                let mut solution = ExtensionSet::new();
+                // Is the solution complete - does this variable *only* refer to itself?
+                let mut solution_complete = true;
+
+                for c in self.get_constraints(&m).unwrap().iter() {
+                    if let Constraint::Plus(r, other_m) = c {
+                        // If `m` has been merged, [`self.variables`] entry
+                        // will have already been updated to the merged
+                        // value by [`self.merge_equal_metas`] so we don't
+                        // need to worry about resolving it.
+                        let other_m = self.resolve(*other_m);
+                        solution.insert(r);
+                        if m != other_m {
+                            // This `m` depends on the solution of `other_m` so
+                            // `solution` is not enough info on its own.
+                            solution_complete = false;
+                            relations.add_edge(m, self.resolve(other_m));
```

Comment:
Do you need to `self.resolve` here? You just did `self.resolve(*other_m)` a moment ago

Reply:
Good point, I was being overly cautious 🙂